src/main/java/com/yahoo/memory/NioBits.java

```diff
@@ -123,46 +120,38 @@ static boolean isPageAligned() {
   }
 
   //RESERVE & UNRESERVE BITS MEMORY TRACKING COUNTERS
-  static void reserveMemory(long capacity) {
-    Util.zeroCheck(capacity, "capacity");
+  static void reserveMemory(final long capacity) {
     try {
-      while (capacity > (1L << 30)) {
-        final long chunk = Math.min(capacity, (1L << 30)); // 1GB chunks
-        NIO_BITS_RESERVE_MEMORY_METHOD.invoke(null, chunk, (int) chunk);
-        capacity -= chunk;
-      }
-
-      if (capacity > 0) {
-        NIO_BITS_RESERVE_MEMORY_METHOD.invoke(null, capacity, (int) capacity);
-      }
-
-      if (isPageAligned) {
-        NIO_BITS_RESERVE_MEMORY_METHOD.invoke(null, pageSize, 0);
-      }
+      reserveUnreserve(capacity, NIO_BITS_RESERVE_MEMORY_METHOD);
     } catch (final Exception e) {
-      throw new RuntimeException("Could not invoke java.nio.Bits.reserveMemory(...)" + e);
+      throw new RuntimeException("Could not invoke java.nio.Bits.reserveMemory(...): "
+          + "capacity = " + capacity + "\n" + e);
     }
   }
 
-  static void unreserveMemory(long capacity) {
-    Util.zeroCheck(capacity, "capacity");
+  static void unreserveMemory(final long capacity) {
     try {
-      while (capacity > (1L << 30)) {
-        final long chunk = Math.min(capacity, (1L << 30)); // 1GB chunks
-        NIO_BITS_UNRESERVE_MEMORY_METHOD.invoke(null, chunk, (int) chunk);
-        capacity -= chunk;
-      }
-      if (capacity > 0) {
-        NIO_BITS_UNRESERVE_MEMORY_METHOD.invoke(null, capacity, (int) capacity);
-      }
-
-      if (isPageAligned) {
-        NIO_BITS_UNRESERVE_MEMORY_METHOD.invoke(null, pageSize, 0);
-      }
-
+      reserveUnreserve(capacity, NIO_BITS_UNRESERVE_MEMORY_METHOD);
     } catch (final Exception e) {
-      throw new RuntimeException("Could not invoke java.nio.Bits.unreserveMemory(...)" + e);
+      throw new RuntimeException("Could not invoke java.nio.Bits.unreserveMemory(...): "
+          + "capacity = " + capacity + "\n" + e);
     }
   }
 
+  private static void reserveUnreserve(long capacity, final Method method) throws Exception {
+    Util.zeroCheck(capacity, "capacity");
+    while (capacity > (1L << 30)) {
+      final long chunk = Math.min(capacity, (1L << 30)); // 1GB chunks
+      method.invoke(null, chunk, (int) chunk);
+      capacity -= chunk;
+    }
+
+    if (capacity > 0) {
+      method.invoke(null, capacity, (int) capacity);
+    }
+
+    if (isPageAligned) {
```

Comment:
Currently this is not needed, because currently we don't actually align. Also it breaks "count" not only when there are allocations larger than 1GB, that are rare, but on every direct allocation, the count is doubled.

Reply:
Fixed.